### PR TITLE
apps wc: Add hierarchical namespace controller

### LINKED ISSUE #1083: [3] Add hierachical namespace controller
**Is your feature request related to a problem? Please describe.**
Based on the [investigation of hierarchical namespaces](https://github.com/elastisys/compliantkubernetes-apps/issues/613) we have decided to add it to compliant kubernetes apps. Adding hierarchical namespace controller (HNC) will let users create namespaces without the assistance of the operations team.

**Describe the solution you'd like**
Create a helm chart for HNC in this repo, based on the manifests in the official HNC release. Preferrably in a HA setup unless a problem is found with that experimental feature. Also add the missing monitoring features (servicemonitor, alerts, grafana dashboard)

If we have enough time, it would be nice to upstream some of this to the official [HNC repo](https://github.com/kubernetes-sigs/hierarchical-namespaces). Alerts and grafana dashboard is probably the easiest part. There is an [open issue](https://github.com/kubernetes-sigs/hierarchical-namespaces/issues/46) for creating a helm chart, but that is asking for a tool that can generate an updated chart each release.

**Describe alternatives you've considered**

**Additional context**

**Definition of done:**
- Helmchart is created and integrated with ck8s-apps
- Monitoring features added
- Public docs updated
- Possibly upstream parts of this to HNC

### PR DESCRIPTION
**Pull request by aarnq**
**What this PR does / why we need it**:
Adds HNC so users can create their own namespaces within namespaces.

**Which issue this PR fixes**:
fixes #1083 

**Public facing documentation PR** *(if applicable)*
elastisys/compliantkubernetes#380

**Special notes for reviewer**:
The RBAC for the controller manager can probably be cut down somewhat, in it's current state it is like upstream.

Also did some fixes for tests that were failing.

**Add a screenshot or an example to illustrate the proposed solution:**
The main metrics are namespace conditions that tells if there are issues with namespaces and subnamespaces, but I added the rest to have as a sort of check to see if it works.
image

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [x] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.
- Chart checklist (pick exactly one):
    - [ ] I upgraded no Chart.
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [ ] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

File: config/config/wc-config.yaml [hunk 1]
```diff
@@ -70,6 +70,64 @@ opa:
     tags:
      - latest
 
+## Hierarchical namespace controller configuration, enable in common-config.yaml
+hnc:
+  ## Enables HA mode for hnc webhooks
+  ha: true
+
+  ## Included namespaces
+  includedNamespacesRegex: ""
```
Comment (OlleLarsson) [inline config/config/wc-config.yaml:79]: You are using `includedNamespaceRegex` in the template
```suggestion
  includedNamespaceRegex: ""
```

File: helmfile/charts/hnc/templates/webhook-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          {{- end }}
+          command:
+            - /manager
+          args:
+            - --webhooks-only
+            - --webhook-server-port=9443
+            - --max-reconciles=10
+            - --apiserver-qps-throttle=50
+            {{- if .Values.includedNamespacesRegex }}
+            - --included-namespace-regex={{ .Values.includedNamespaceRegex }}
```
Comment (OlleLarsson) [inline helmfile/charts/hnc/templates/webhook-deployment.yaml:53]: Mismatch here
```suggestion
            {{- if .Values.includedNamespacesRegex }}
            - --included-namespace-regex={{ .Values.includedNamespacesRegex }}
```

File: helmfile/charts/hnc/templates/manager-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            {{- if not .Values.webhookDeployment.create }}
+            - --webhook-server-port=9443
+            {{- else }}
+            - --no-webhooks
+            {{- end }}
+            - --metrics-addr=:8080
+            - --max-reconciles=10
+            - --apiserver-qps-throttle=50
+            {{- if .Values.includedNamespacesRegex }}
+            - --included-namespace-regex={{ .Values.includedNamespaceRegex }}
```
Comment (OlleLarsson) [inline helmfile/charts/hnc/templates/manager-deployment.yaml:57]: Same
```suggestion
            {{- if .Values.includedNamespacesRegex }}
            - --included-namespace-regex={{ .Values.includedNamespacesRegex }}
```

File: helmfile/values/hnc.yaml.gotmpl [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  - kube-system
+  - kured
+  - monitoring
+  - postgres-system
+  - rabbitmq-system
+  - redis-system
+  - rook-ceph
+  - velero
+  {{- with .Values.hnc.excludedNamespaces }}
+  {{- toYaml . | nindent  }}
```
Comment (OlleLarsson) [inline helmfile/values/hnc.yaml.gotmpl:21]:
```suggestion
  {{- toYaml . | nindent 2 }}
```

**Review by OlleLarsson — COMMENTED** [review]
I have tested basic functionality and it seems to work as expected and intended.
Anything in particular that you want me to look at or try out?

**Comment by aarnq** [comment]
> I have tested basic functionality and it seems to work as expected and intended. Anything in particular that you want me to look at or try out?

Not really, as long as the basics work and you haven't had any issues with the webhooks then it should be fine.

I'm gonna run some tests on my end to ensure that the webhooks doesn't interfere when applying apps, it did so for me earlier but it might be due to it tagging almost every namespace the first time I deployed it.

File: helmfile/charts/hnc/templates/user-rbac.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+- apiGroups:
+  - hnc.x-k8s.io
+  resources:
+  - hierarchyconfigurations
+  verbs:
+  - get
+  - list
+  - patch
+  - update
+{{- end }}
```
Comment (viktor-f) [inline helmfile/charts/hnc/templates/user-rbac.yaml:22]: I suggest that they should always be able to `get,list,watch` the hierachyconfig, maybe even the cluster config `hncconfiguration`.

File: config/config/wc-config.yaml (hunk 1 above)
Comment (viktor-f) [inline config/config/wc-config.yaml:79]: Are the user namespaces included by default?

File: helmfile/charts/hnc/values.yaml [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+webhookService:
+  create: true
+  port: 443
+
+certificate:
+  create: true
+
+serviceMonitor:
+  create: true
+  relabelings: []
```
Comment (viktor-f) [inline helmfile/charts/hnc/values.yaml:112]: More config is often good. But would you ever want to not deploy manager-*, webhook, and webhookservice (plus maybe some others)? I'm thinking that the `create` options could be removed from some.
But maybe I'm wrong, so please let me know your thoughts on this.

File: helmfile/charts/prometheus-alerts/values.yaml [hunk 7]
```diff
@@ -27,6 +27,7 @@ defaultRules:
     certManagerCertificateAlerts: true
     falcoAlerts: true
     general: true
+    hnc: false
```
Comment (viktor-f) [inline helmfile/charts/prometheus-alerts/values.yaml:30]: Do we not want this turned on if we enable hnc?

**Comment by viktor-f** [comment]
Regarding the dashboard. Will the green texts under "conditions" turn red if one of those errors actually happen?

File: config/config/wc-config.yaml [hunk 8]
```diff
@@ -70,6 +70,64 @@ opa:
...
+  ## Labels that will be propagated to subnamespaces (allows regex)
+  ## Labels in particular must also be configured in the HierarchyConfiguration object to be propagated
+  managedNamespaceLabels: []
+
+  ## Set to allow users to edit the HierarchyConfiguration in each (sub)namespace
+  allowConfigEdit: false
+
+  ## Manager deployment configuration
+  manager:
+    replicaCount: 1
```
[the remark this answers is not in the document]
Comment (aarnq) [reply config/config/wc-config.yaml:98]: Leftover from the template generated by helm create. Should I remove it?

[the remark this answers is not in the document]
Comment (OlleLarsson) [reply config/config/wc-config.yaml:98]: Yep

File: helmfile/charts/hnc/values.yaml (hunk 6 above)
Comment (aarnq) [reply helmfile/charts/hnc/values.yaml:112]: It's mostly a leftover from when I created the chart and tied it to the build process in the HNC repo. I often compared the result to their generated manifests so it was easier to do so when I could select which one to template.

File: helmfile/charts/prometheus-alerts/values.yaml (hunk 7 above)
Comment (aarnq) [reply helmfile/charts/prometheus-alerts/values.yaml:30]: Yes, and that is set up in the values file `prometheus-alerts-sc.yaml.gotmpl` but I forgot to do the same with the Grafana dashboard.

**Comment by aarnq** [comment]
> Regarding the dashboard. Will the green texts under "conditions" turn red if one of those errors actually happen?

Yes, I'll add a text or number of conditions as well.

File: helmfile/charts/prometheus-alerts/values.yaml (hunk 7 above)
Comment (viktor-f) [reply helmfile/charts/prometheus-alerts/values.yaml:30]: Right, I mixed up the values files.

File: helmfile/charts/hnc/values.yaml (hunk 6 above)
Comment (aarnq) [reply helmfile/charts/hnc/values.yaml]: I've simplified the chart a bit now and added the numbers to the dashboard.

**Review by viktor-f — APPROVED** [review]
LGTM

**Review by crssnd — APPROVED** [review]

**Review by lucianvlad — APPROVED** [review]
LGTM
Fix the conflicts
Also merge the commits into 1 commit if possible .

**Pull request merged**